revocation/crl/cache/memory.go

```diff
@@ -0,0 +1,76 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// MemoryCache is an in-memory cache that stores CRL bundles.
+//
+// The cache is built on top of the sync.Map to leverage the concurrency control
+// and atomicity of the map, so it is suitable for writing once and reading many
+// times. The CRL is stored in memory as a Bundle type.
+//
+// MemoryCache doesn't handle cache cleaning but provides the Delete and Clear
+// methods to remove the CRLs from the memory.
+type MemoryCache struct {
+	// MaxAge is the maximum age of the CRLs cache. If the CRL is older than
+	// MaxAge, it will be considered as expired.
+	MaxAge time.Duration
+
+	store sync.Map
+}
+
+// NewMemoryCache creates a new memory store.
+func NewMemoryCache() *MemoryCache {
+	return &MemoryCache{
+		MaxAge: DefaultMaxAge,
+	}
+}
+
+// Get retrieves the CRL from the memory store.
+//
+// - if the key does not exist, return ErrNotFound
+// - if the CRL is expired, return ErrCacheMiss
+func (c *MemoryCache) Get(ctx context.Context, uri string) (*Bundle, error) {
+	value, ok := c.store.Load(uri)
+	if !ok {
+		return nil, ErrCacheMiss
+	}
+	bundle, ok := value.(*Bundle)
+	if !ok {
+		return nil, fmt.Errorf("invalid type: %T", value)
+	}
+
+	expires := bundle.Metadata.CachedAt.Add(c.MaxAge)
+	if c.MaxAge > 0 && time.Now().After(expires) {
+		return nil, ErrCacheMiss
+	}
```

Comment:
How are we doing housekeeping i.e. removing/purging the expired items in cache?

Reply:
We will replace the old one with the new one the next time we hit the same cache. There is no active logic to delete expired ones now. Maybe it can be an improvement to provide a cleaning solution in the future.

Reply:
Moved `MemoryCache` to be an internal `cachehelper` package for testing purposes. And removed `FileCache`

Reply:
Can we use some dependency which provides caching solution ?

Reply:
Yes, there are many cache libraries, but they need to be wrapped to the CRL Cache interface, which is easy. For example, https://github.com/eko/gocache.

I think we should limit the external dependencies in core-go to avoid security issues. Core-go users can use external cache libraries.

Reply:
Also, the `MemoryCache` is only for internal unit test now.